antenna/ext/gcs/crashstorage.py

```diff
@@ -0,0 +1,180 @@
+
+        $ gcloud auth application-default login
+
+    - If you're running your application elsewhere, you should download a `service account
+      <https://cloud.google.com/iam/docs/creating-managing-service-accounts#creating>`_
+      JSON keyfile and point to it using an environment variable::
+
+        $ export GOOGLE_APPLICATION_CREDENTIALS="/path/to/keyfile.json"
+
+
+    **Local emulator**
+
+    If you set the environment variable ``STORAGE_EMULATOR_HOST=http://host:port``,
+    then this will connect to a local GCS emulator.
+
+
+    """
+
+    class Config:
+        bucket_name = Option(
+            doc=(
+                "Google Cloud Storage bucket to save to. Note that the bucket must "
+                "already have been created."
+            ),
+        )
+
+    def __init__(self, config):
+        self.config = config.with_options(self)
+        self.bucket = self.config("bucket_name")
+
+        if emulator := os.environ.get("STORAGE_EMULATOR_HOST"):
+            logger.debug(
+                "STORAGE_EMULATOR_HOST detected, connecting to emulator: %s",
+                emulator,
+            )
+            self.client = storage.Client(
+                credentials=AnonymousCredentials(),
+                project="test",
+            )
+        else:
+            self.client = storage.Client()
+
+        register_for_verification(self.verify_write_to_bucket)
+
+    def _save_file(self, path, data):
+        """Save a single file to GCS.
+
+        :arg str path: the path to save to
+        :arg bytes data: the data to save
+
+        """
+        bucket = self.client.get_bucket(self.bucket)
+        blob = bucket.blob(path)
+        blob.upload_from_string(data)
+
+    def verify_write_to_bucket(self):
+        """Verify GCS bucket exists and can be written to."""
+        self._save_file(generate_test_filepath(), b"test")
+
+    def check_health(self, state):
+        """Check GCS connection health."""
+        try:
+            # get the bucket to verify GCS is up and we can connect to it.
+            self.client.get_bucket(self.bucket)
+        except Exception as exc:
+            state.add_error("GcsCrashStorage", repr(exc))
+
+    def _get_raw_crash_path(self, crash_id):
+        return "v1/raw_crash/{date}/{crash_id}".format(
+            date=get_date_from_crash_id(crash_id),
+            crash_id=crash_id,
+        )
+
+    def _get_dump_names_path(self, crash_id):
+        return f"v1/dump_names/{crash_id}"
+
+    def _get_dump_name_path(self, crash_id, dump_name):
+        if dump_name in (None, ""):
+            dump_name = "dump"
+
+        return "v1/{dump_name}/{crash_id}".format(
+            dump_name=dump_name, crash_id=crash_id
+        )
+
+    def save_raw_crash(self, crash_id, raw_crash):
+        """Save the raw crash and related dumps.
+
+        .. Note::
+
+           If you're saving the raw crash and dumps, make sure to save the raw
+           crash last.
```

Comment:
When/if we get rid of the AWS Lambda job that triggers on object save events for raw crash files in S3 and don't replace it with an equivalently implemented thing in GCP, then we can drop this requirement.

Reply:
sounds good, i plan to do that before cutting over prod, so i'll remove this comment and the ones later about calling save_dumps first